### PR TITLE
create project model with test cases and associations

### PR DESCRIPTION
**Pull request by HMTanzeel**

### CODE REVIEW

File: app/models/project.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class Project < ApplicationRecord
+  belongs_to :user
+  has_many :tasks
```
Comment (aliammaar) [inline app/models/project.rb]: add ` dependent: :destroy`

File: config/database.yml [hunk 2]
```diff
@@ -7,3 +7,11 @@ development:
   database: task_manager_development
   username: postgres
   password: postgres
+
+test:
+  <<: *default
+  adapter: postgresql
+  database: task_manager_development
```
Comment (aliammaar) [inline config/database.yml]: why you are using same database for development and test env ?

File: test/models/project_test.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+require "test_helper"
+
+class ProjectTest < ActiveSupport::TestCase
+  test "valid with user" do
+    user = User.create(name: "Ali", mobile: "0900", email: "a2@g.com", password_digest: "123", address: "ajhsncjfacnaje")
```
Comment (aliammaar) [inline test/models/project_test.rb]: Are we writing test for Project ? I am not seeing anything being validated on project model in this test case.

File: test/models/project_test.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    assert user.valid?
+
+    assert_empty user.errors
+  end
+
+  test "invalid without user" do
+    user = User.create()
+    assert user.valid?
```
Comment (aliammaar) [inline test/models/project_test.rb]: same as above, you are testing user not project.

File: test/models/project_test.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  test "invalid with user and if title is empty" do
+    user = User.create(name: "Ali", mobile: "0900", email: "a2@g.com", password_digest: "123", address: "ajhsncjfacnaje")
+    project = Project.new
+    project.user = user
+    assert project.valid?
+
+    assert_empty project.errors
+  end
+
+  test "invalid if title and user is not available" do
```
Comment (aliammaar) [inline test/models/project_test.rb]: change to: `is not valid without title and user`

File: app/models/project.rb (hunk 1 above)
Comment (HMTanzeel) [reply app/models/project.rb]: done

File: config/database.yml (hunk 2 above)
Comment (HMTanzeel) [reply config/database.yml]: done

File: test/models/project_test.rb (hunk 3 above)
Comment (HMTanzeel) [reply test/models/project_test.rb]: done

File: test/models/project_test.rb (hunk 4 above)
Comment (HMTanzeel) [reply test/models/project_test.rb]: done

File: test/models/project_test.rb (hunk 5 above)
Comment (HMTanzeel) [reply test/models/project_test.rb]: done

File: config/database.yml [hunk 6]
```diff
@@ -7,3 +7,11 @@ development:
...
   username: postgres
   password: postgres
+
+test:
+  <<: *default
+  adapter: postgresql
+  database: task_manager_test
+  username: postgres
+  password: postgres 
+
```
Comment (aliammaar) [inline config/database.yml]: remove extra newline.

File: test/models/project_test.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    user = User.create(name: "Ali", mobile: "0900", email: "a2@g.com", password_digest: "123", address: "ajhsncjfacnaje")
+    project = Project.new(title: "any")
+    project.user = user
+    assert project.valid?
+
+    assert_empty project.errors
+  end
+
+  test "not valid if user is available and title is empty" do
+    user = User.create(name: "Ali", mobile: "0900", email: "a2@g.com", password_digest: "123", address: "ajhsncjfacnaje")
```
Comment (aliammaar) [inline test/models/project_test.rb]: use fixture for the user model

File: config/database.yml (hunk 6 above)
Comment (HMTanzeel) [reply config/database.yml]: done

File: test/models/project_test.rb (hunk 7 above)
Comment (HMTanzeel) [reply test/models/project_test.rb]: done

File: app/views/projects/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+<%= form_with model: [current_user, @project] do |form| %>
+  <form class='mb-5'>
+    <div class='container'>
+      <div class='card-header'>
+        <h4></h4>
```
Comment (aliammaar) [inline app/views/projects/_form.html.erb:5]: ?

File: test/fixtures/users.yml [hunk 9]
```diff
@@ -1,15 +1,15 @@
...
 
 two:
-  name: MyString
-  mobile: 1
-  email: MyString
-  password_digest: <%= BCrypt::Password.create("secret") %>
-  address: MyString
+  name: Ali
+  mobile: 5678
+  email: a2@g.com
```
Comment (aliammaar) [inline test/fixtures/users.yml]: user proper dummy attributes!

File: spec/factories/users.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :user do
+    name { 'Tanzeel' }
+    mobile { '1234' }
+    email { 'a1@g.com' }
```
Comment (aliammaar) [inline spec/factories/users.rb]: an appropriate email!

File: spec/factories/users.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+    password_digest { '12345' }
+    address { 'qwertyuiop' }
```
Comment (aliammaar) [inline spec/factories/users.rb]: Lahore, Pakistan

File: spec/factories/users.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :user do
+    name { 'Tanzeel' }
+    mobile { '1234' }
+    email { 'a1@g.com' }
+    password_digest { '12345' }
```
Comment (aliammaar) [inline spec/factories/users.rb]: Password1234

File: test/fixtures/projects.yml [hunk 13]
```diff
@@ -0,0 +1,21 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  title: E-Commerece
+  description: Cart,Checkout,Proceed,Paynow
```
Comment (aliammaar) [inline test/fixtures/projects.yml]:
```
Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod
tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

File: spec/models/project_spec.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe Project, type: :model do
+  it "valid if title is available" do
+    user = FactoryBot.create(:user)
```
Comment (aliammaar) [inline spec/models/project_spec.rb]: try create(:user)

File: spec/models/project_spec.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+    project = FactoryBot.create(:project)
+    project.user = user
+    expect(project.title).present?
```
Comment (aliammaar) [inline spec/models/project_spec.rb:16]: use equal to match title

File: spec/models/user_spec.rb [hunk 16]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  it "create a user" do
+    user1 = FactoryBot.create(:user)
+    expect(user1).to be_valid
```
Comment (aliammaar) [inline spec/models/user_spec.rb]: it is User.count to be 1

File: spec/factories/projects.rb [hunk 17]
```diff
@@ -0,0 +1,11 @@
+FactoryBot.define do
+  factory :project do
+    title { 'E-Commerece' }
+    description { 'Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.' }
+    active { false }
+    started_at { '2022-03-11 12:03:27' }
+    close_at { '2022-03-11 12:03:27' }
+    project_type { 1 }
+    user_id { 1 }
```
Comment (aliammaar) [inline spec/factories/projects.rb]: add association through factory

File: spec/factories/users.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :user do
+    id { 1 }
```
Comment (aliammaar) [inline spec/factories/users.rb]: never add id in spec

File: spec/factories/users.rb [hunk 19]
```diff
@@ -0,0 +1,10 @@
...
+    name { 'Tanzeel' }
+    mobile { '1234' }
+    email { 'mtanzeel6633@gmail.com' }
+    password_digest { 'Password1234' }
+    address { 'Lahore, Pakistan' }
```
Comment (aliammaar) [inline spec/factories/users.rb]: only add required attributes

File: spec/models/user_spec.rb [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+  end
+
+  it "valid if name is equal" do
```
Comment (aliammaar) [inline spec/models/user_spec.rb]: meaningless

File: spec/models/user_spec.rb [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+    user1 = FactoryBot.create(:user)
+    expect(user1.name).to eq('Tanzeel')
+  end
+
+  it "valid if id is equal" do
```
Comment (aliammaar) [inline spec/models/user_spec.rb]: meaningless

**Review by aliammaar — CHANGES_REQUESTED** [review]

**Review by aliammaar — APPROVED** [review]

**Pull request merged**